qutip/core/cy/qobjevo.pyx

```diff
@@ -1017,3 +1017,32 @@ cdef class QobjEvo:
             part = (<_BaseElement> element)
             out = part.matmul_data_t(t, state, out)
         return out
+
+
+cdef class QobjEvoHerm(QobjEvo):
+    """
+    QobjEvo with matmul_data variant that use the ``herm_matmul`` special case.
+    """
+    def __init__(self, Q_object, args=None, tlist=None,
+                 order=3, copy=True, compress=True,
+                 function_style=None, boundary_conditions=None):
+        super().__init__(
+            Q_object, args, tlist, order, copy,
+            compress, function_style, boundary_conditions
+        )
+        subsize = int(self.shape[1] ** 0.5)
+
+    cpdef Data matmul_data(QobjEvoHerm self, object t, Data state, Data out=None):
+        """Compute ``out += self(t) @ state``"""
+        cdef _BaseElement part
+        t = self._prepare(t, state)
+        if out is None and type(state) is Dense:
+            out = dense.zeros(self.shape[0], state.shape[1],
+                      (<Dense> state).fortran)
+        elif out is None:
+            out = _data.zeros[type(state)](self.shape[0], state.shape[1])
+
+        for element in self.elements:
+            part = (<_BaseElement> element)
+            out = part.matmul_data_t_herm(t, state, self.subsize, out)
```

Comment:
This requires each matrix element of the `QobjEvo` to be Hermitian, which is neither necessary not sufficient for the `qobjevo(t)` to be Hermitian.

On the up side, it is fairly easy to check whether each matrix element in a QobjEvo is hermitian, so perhaps we could have each element use the hermitian version automatically if it knows its Qobj is hermitian?

Reply:
This is not meant to be used when `qobjevo(t)` to be Hermitian. 
The output of `vector_to_operator(qobjevo(t) @ operator_to_vector(rho))` should be hermitian.
Since the `out` is only accumulated on, it does not need to be true for every element independently, just the result.